provider/src/main/java/com/synopsys/integration/alert/provider/blackduck/task/accumulator/BlackDuckNotificationRetriever.java

```diff
@@ -76,52 +79,21 @@ private String toDateString(SimpleDateFormat sdf, OffsetDateTime offsetDateTime)
         return sdf.format(date);
     }
 
-    public static class BlackDuckNotificationPage {
-        private final int currentOffset;
-        private final int currentLimit;
-        private final int currentTotal;
-        private final List<NotificationView> currentNotifications;
-        private final NotificationPageRetriever notificationPageRetriever;
-
-        private BlackDuckNotificationPage(
-            int currentOffset,
-            int currentLimit,
-            int currentTotal,
-            List<NotificationView> currentNotifications,
-            NotificationPageRetriever notificationPageRetriever
-        ) {
-            this.currentOffset = currentOffset;
-            this.currentLimit = currentLimit;
-            this.currentTotal = currentTotal;
-            this.currentNotifications = currentNotifications;
-            this.notificationPageRetriever = notificationPageRetriever;
-        }
+    private class NextNotificationPageRetriever implements NextPageRetriever<NotificationView, IntegrationException> {
 
-        public BlackDuckNotificationPage retrieveNextPage() throws IntegrationException {
-            int newOffset = currentOffset + currentLimit;
-            if (hasNextPage()) {
-                BlackDuckPageDefinition pageDefinition = new BlackDuckPageDefinition(currentLimit, newOffset);
-                BlackDuckPageResponse<NotificationView> notificationPage = notificationPageRetriever.apply(pageDefinition);
-                return new BlackDuckNotificationPage(newOffset, currentLimit, notificationPage.getTotalCount(), notificationPage.getItems(), notificationPageRetriever);
-            }
-            return new BlackDuckNotificationPage(newOffset, currentLimit, 0, List.of(), notificationPageRetriever);
-        }
-
-        public List<NotificationView> getCurrentNotifications() {
-            return currentNotifications;
-        }
+        private final BlackDuckRequestBuilder blackDuckRequestBuilder;
 
-        public boolean isEmpty() {
-            return currentTotal == 0;
+        public NextNotificationPageRetriever(BlackDuckRequestBuilder blackDuckRequestBuilder) {
+            this.blackDuckRequestBuilder = blackDuckRequestBuilder;
         }
 
-        private boolean hasNextPage() {
-            return currentTotal == currentLimit;
+        @Override
+        public AlertPagedDetails<NotificationView> retrieveNextPage(int currentOffset, int currentLimit) throws IntegrationException {
+            int newOffset = currentOffset + currentLimit;
+            BlackDuckPageDefinition pageDefinition = new BlackDuckPageDefinition(currentLimit, newOffset);
+            BlackDuckPageResponse<NotificationView> notificationPage = retrievePageOfFilteredNotifications(blackDuckRequestBuilder, pageDefinition);
+            return new AlertPagedDetails(newOffset, currentLimit, notificationPage.getTotalCount(), notificationPage.getItems());
```

Comment:
Seeing this implementation makes me think the interface should be called not be a "next" page retriever, but simply a page retriever. Then the increment logic can be in `StatefulAlertPage`. Otherwise every implementation will repeat the step of `int newOffset = currentOffset + currentLimit;` which is the thing I believe we are trying to abstract away from the end-user of `StatefulAlertPage`.